### PR TITLE
Tool to add a top layer with specified thickness to a mesh 

### PR DESCRIPTION
**Pull request by TomFischer**
Erik and Marc requested a small tool to add a new top layer to a mesh. Hope it will be useful.
See also the [documentation](http://docs.opengeosys.org/docs/tools/meshing/addtoplayer).

### CODE REVIEW

File: Applications/Utils/MeshEdit/AddTopLayer.cpp [hunk 1]
```diff
@@ -0,0 +1,175 @@
+/*
+ * \date 2015-04-14
+ * \brief Adds a top layer to an existing mesh.
+ */
```
Comment (endJunction) [inline Applications/Utils/MeshEdit/AddTopLayer.cpp]: copyright header missing.

File: Applications/Utils/MeshEdit/AddTopLayer.cpp [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+	std::vector<MeshLib::Element*> & subsfc_elements(
+		const_cast<std::vector<MeshLib::Element*> &>(subsfc_mesh->getElements())
+	);
+	std::size_t orig_size(subsfc_elements.size());
+	std::vector<MeshLib::Element*> const& sfc_elements(sfc_mesh->getElements());
+	std::size_t const n_sfc_elements(sfc_elements.size());
+	for (std::size_t k(0); k<n_sfc_elements; ++k) {
+		MeshLib::Element const*const sfc_elem(sfc_elements[k]);
+		if (sfc_elem->getGeomType() == MeshElemType::TRIANGLE) {
+			// add a prism
```
Comment (endJunction) [inline Applications/Utils/MeshEdit/AddTopLayer.cpp]: I'd extract the innermost if into own function.

Is an element of POINT type not implemented or is it just missing here?

**Comment by endJunction** [comment]
:fast_forward: after the copyright header is corrected.

**Comment by bilke** [comment]
Nice docs!

File: Applications/Utils/MeshEdit/AddTopLayer.cpp (hunk 2 above)
Comment (TomFischer) [inline Applications/Utils/MeshEdit/AddTopLayer.cpp]: > I'd extract the content of the innermost if into own function.

I'll try to extract the instructions of the innermost if into separate methods.

> Is an element of POINT type not implemented or is it just missing here?

At the moment we have not an element type POINT.

File: Applications/Utils/MeshEdit/AddTopLayer.cpp [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+// GeoLib
+#include "GeoLib/PointWithID.h"
+
+// MeshLib
+#include "MeshLib/Mesh.h"
+#include "MeshLib/Node.h"
+#include "MeshLib/Elements/Element.h"
+#include "MeshLib/Elements/Prism.h"
+#include "MeshLib/Elements/Hex.h"
+#include "MeshLib/Elements/Quad.h"
```
Comment (rinkk) [inline Applications/Utils/MeshEdit/AddTopLayer.cpp]: technically, several includes are missing (tri, line, etc.)

File: Applications/Utils/MeshEdit/AddTopLayer.cpp [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+// ThirdParty/logog
+#include "logog/include/logog.hpp"
+
+// BaseLib
+#include "BaseLib/LogogSimpleFormatter.h"
+#include "BaseLib/FileTools.h"
+
+// FileIO
+#include "FileIO/readMeshFromFile.h"
+#include "FileIO/XmlIO/Boost/BoostXmlGmlInterface.h"
```
Comment (rinkk) [inline Applications/Utils/MeshEdit/AddTopLayer.cpp]: include not needed

File: Applications/Utils/MeshEdit/AddTopLayer.cpp [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+#include "FileIO/VtkIO/VtuInterface.h"
+
+// GeoLib
+#include "GeoLib/PointWithID.h"
```
Comment (rinkk) [inline Applications/Utils/MeshEdit/AddTopLayer.cpp]: include not needed

File: Applications/Utils/MeshEdit/AddTopLayer.cpp [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+	logog::Cout* logog_cout (new logog::Cout);
+	BaseLib::LogogSimpleFormatter *custom_format (new BaseLib::LogogSimpleFormatter);
+	logog_cout->SetFormatter(*custom_format);
+
+	TCLAP::CmdLine cmd(
+		"Adds a top layer to an existing 3d mesh",
+		' ',
+		"0.1");
+
+	TCLAP::ValueArg<std::string> mesh_arg("m", "mesh-file",
```
Comment (rinkk) [inline Applications/Utils/MeshEdit/AddTopLayer.cpp]: to be consistent with all the other tools you could write ```mesh_in_arg("i", "input-mesh-file",....)```

File: Applications/Utils/MeshEdit/AddTopLayer.cpp [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+	INFO("Extracting top surface of mesh \"%s\" ... ",
+		mesh_arg.getValue().c_str());
+	const MathLib::Vector3 dir(0,0,-1);
+	double const angle(90);
+	std::unique_ptr<MeshLib::Mesh> sfc_mesh(
+		MeshLib::MeshSurfaceExtraction::getMeshSurface(
+			*subsfc_mesh, dir, angle, true
+		)
+	);
+	INFO("done.");
```
Comment (rinkk) [inline Applications/Utils/MeshEdit/AddTopLayer.cpp]: you are not testing if the input mesh is actually 3D (not that it matters, the method will work just as well with a 2D mesh ... it is just not consistent with the documentation in the beginning)

File: Applications/Utils/MeshEdit/AddTopLayer.cpp [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+	);
+	std::size_t const n_subsfc_nodes(subsfc_nodes.size());
+
+	std::vector<MeshLib::Node*> const& sfc_nodes(sfc_mesh->getNodes());
+	std::size_t const n_sfc_nodes(sfc_nodes.size());
+
+	double const layer_thickness(layer_thickness_arg.getValue());
+	// *** copy sfc nodes to subsfc mesh node
+	std::map<std::size_t, std::size_t> subsfc_sfc_id_map;
+	for (std::size_t k(0); k<n_sfc_nodes; ++k) {
```
Comment (rinkk) [inline Applications/Utils/MeshEdit/AddTopLayer.cpp]: for a deep-copy of nodes/elements you can use the methods from ```DuplicateMeshComponents```, that will save you some code here ...

Comment (TomFischer) [inline Applications/Utils/MeshEdit/AddTopLayer.cpp]: In this loop simultaneously the `subsfc_sfc_id_map` is filled and the z-coordinate is modified.

File: Applications/Utils/MeshEdit/AddTopLayer.cpp (hunk 7 above)
Comment (TomFischer) [inline Applications/Utils/MeshEdit/AddTopLayer.cpp]: Generalized the documentation from "3d mesh" to "mesh".

**Pull request merged**